### ISSUE TITLE
About the on-chip cache

### ISSUE THREAD

**Issue by doonny** [zohourih/FPGAMemBench#8]
Hi, it is great to see this interesting work.  

I have a small question on measuring the efficiency of the memory controller,  have you considered disabling the generated on-chip cache ?   when the halo size is larger than zero, which means that a small amount of the overlapped region between adjacent blocks will be reused by the kernel, which might be stored in the cache generated by the off-line compiler.  This might improves the efficiency of the external memory bandwidth, but it is not achieved by the DDR controller.

**Comment by zohourih** [comment]
@doonny Thank you for your interest in the work. In fact the cache was disabled (using compiler flag rather than volatile) for all of our experiments and the results reported in the paper. Though we did also experiment with the cache enabled, the results of which were reported in the subsection called "Effect of Cache" on the paper. Essentially, the cache was not generated at all in any of the kernels with the newer versions of the compiler, and older versions only implemented it for single work-item kernels which had no noticeable positive or negative effect on the performance. If I remember correctly, the cache has a depth of 512 or 1024 indexes which is considerably smaller than our tile sizes and uses a very simple mechanism; I believe all its data get replaced by the time we move from one tile to the next one anyway.

**Comment by doonny** [comment]
Hi, thanks, I will check the paper for the details.

One more question on the code,  you are using the following statement in computing the address:

x = (x + VEC) & (BLOCK_X - 1);

Does (BLOCK_X-1) serves as a address MASK to restrict x in the range of [0 bsize-1] ?

**Comment by zohourih** [comment]
Yes, that expression is equal to `x = (x + VEC) % BLOCK_X` or:

```
x += VEC;
if (x == BLOCK_X)
{
    x = 0;
}
```

The expression I have used is just simple bit masking and hence, results in a shorter combination path and better operating frequency compared to the cases above, since the chain of updates on the co-ordinate variables is the critical path of the design, but it only works if BLOCK_X is a power of two and divisible by VEC.

**Comment by doonny** [comment]
Hi zohourih,  Thx for the reply. 

I have made a fewer quick tests on my DE5a_net_ddr4 board by using SDK v20.1, and It is found that data alignment (padding) does not have obvious impact on the BW anymore. Here's some dumped result with pad=3 and 0:

================================================================================
Selected platform number: 0
Device count: 2
Device type: 8
Selected device: de5a_net_ddr4 : Arria 10 Reference Platform (aclde5a_net_ddr40)
================================================================================

Kernel type:           Standard
Kernel model:          Single Work-item
Array size:            26214400 indexes
Buffer size:           100 MiB
Total memory usage:    400 MiB
Block size:            1024
Vector size:           16
Array padding:         3
Halo width             0

Creating host buffers...
Filling host buffers with random data...
Creating device buffers...
Writing data to device...
Device warm-up...
Executing "R1W0" kernel...
No verification available for this kernel!
Executing "R1W1" kernel...
Reading data back from device...
Verifying "R1W1" kernel: SUCCESS!
Executing "R2W1" kernel...
Reading data back from device...
Verifying "R2W1" kernel: SUCCESS!
Executing "R3W1" kernel...
Reading data back from device...
Verifying "R3W1" kernel: SUCCESS!
Executing "R2W2" kernel...
Reading data back from device...
Verifying "R2W2" kernel: SUCCESS!

Redundancy: 0.00%
R1W0: 18.176 GB/s (16.928 GiB/s) @5.8 ms
R1W1: 22.762 GB/s (21.199 GiB/s) @9.2 ms
R2W1: 23.313 GB/s (21.711 GiB/s) @13.5 ms
R3W1: 24.376 GB/s (22.702 GiB/s) @17.2 ms
R2W2: 21.554 GB/s (20.074 GiB/s) @19.5 ms


================================================================================
Selected platform number: 0
Device count: 2
Device type: 8
Selected device: de5a_net_ddr4 : Arria 10 Reference Platform (aclde5a_net_ddr40)
================================================================================

Kernel type:           Standard
Kernel model:          Single Work-item
Array size:            26214400 indexes
Buffer size:           100 MiB
Total memory usage:    400 MiB
Block size:            1024
Vector size:           16
Array padding:         0
Halo width             0

Creating host buffers...
Filling host buffers with random data...
Creating device buffers...
Writing data to device...
Device warm-up...
Executing "R1W0" kernel...
No verification available for this kernel!
Executing "R1W1" kernel...
Reading data back from device...
Verifying "R1W1" kernel: SUCCESS!
Executing "R2W1" kernel...
Reading data back from device...
Verifying "R2W1" kernel: SUCCESS!
Executing "R3W1" kernel...
Reading data back from device...
Verifying "R3W1" kernel: SUCCESS!
Executing "R2W2" kernel...
Reading data back from device...
Verifying "R2W2" kernel: SUCCESS!

Redundancy: 0.00%
R1W0: 18.237 GB/s (16.985 GiB/s) @5.7 ms
R1W1: 22.707 GB/s (21.148 GiB/s) @9.2 ms
R2W1: 23.578 GB/s (21.959 GiB/s) @13.3 ms
R3W1: 24.590 GB/s (22.902 GiB/s) @17.1 ms
R2W2: 21.653 GB/s (20.166 GiB/s) @19.4 ms

-------------------------------------------------------------------------
Do you think Intel have changed the DDR controller design or I am not making things correctly ?

**Comment by zohourih** [comment]
I find it very unlikely for them to have made any changes in the memory interface. I believe you might be oversubscribing the memory controller and hence, the effect of memory alignment is not showing up. What is the operating frequency of the kernel in your case? Also, did you disable interleaving? Maybe try with a vector size of 8 and interleaving disabled to see if there is any difference. If not, you can try older versions of the compiler to see if there is any difference and they have actually made changes in the new versions. Unfortunately my Nallatech board never got newer BSPs and I could not test anything beyond v18.1 of the compiler.

**Comment by doonny** [comment]
I am using the following configuration :

make std INTEL_FPGA=1 BOARD=~/intelFPGA_pro/BSP/de5a_net_ddr4  SEED=3 VEC=16 NO_INTER=1 TFMAX=266

and for host:     make std INTEL_FPGA=1 HOST_ONLY=1  NO_INTER=1 VEC=16

Is this configuration OK ?

BTW, could you explain a bit more about why oversubscribing memory controller would affect the behavior of alignment ?

**Comment by zohourih** [comment]
Yes, that should disable interleaving. However, TFMAX will just change the target operating frequency of the compiler and can potentially allow achieving higher operating frequency by inserting more registers into the pipeline. If you want to override the post-place-and-route frequency, then you have to apply the hack mentioned in the README and use `FMAX=VALUE`. Your board seems to have 2400 MHz DDR4 memory and the memory controller runs at 300 MHz. To saturate the memory bandwidth, the kernel should run at 300 MHz and a vector size of 16 with interleaving disabled.

In my experience, if the memory controller is oversusbscribed, e.g. using a vector size of 32 and the same operating frequency as the memory controller, then the effect of alignment will become quite small (it is still there, though). In your case, I thought you might not have disabled interleaving in which case the achieved bandwidth would hardly go beyond 60% of peak and your measured performance could already be at oversubscription state; however, if you have disabled interleaving, then this should not happen until you reach something close to 38 GB/s.
